### ISSUE TITLE
stklos with no flags to run a script?

### ISSUE THREAD

**Issue by lassik** [egallesio/STklos#432]
Given a file `hello.scm`:

```
(display "Hello world")
(newline)
```

This runs the script:

```
$ stklos-script hello.scm
Hello world
```

But this doesn't:

```
$ stklos hello.scm
  \    STklos version 1.70  (Id: eeb9977d)
   \   Copyright (C) 1999-2022 Erick Gallesio <eg@unice.fr>
  / \  Université Côte d'Azur
 /   \ [Darwin-19.6.0-x86_64/pthread/no-readline/utf8]
stklos> (command-name)
#f
stklos> (command-args)
("hello.scm")
stklos> (script-file)
#f
stklos>
```

Since the latter does not seem to do anything useful with the command line arguments, and many other interpreters' main commands run scripts (e.g. `gosh hello.scm`, `chibi-scheme hello.scm`, `python hello.py`, `ruby hello.rb`), could `stklos hello.scm` run a script as well?

**Comment by jpellegrini** [comment]
`stklos-script hello.scm` does what you want, I think...

**Comment by egallesio** [comment]
> `stklos-script hello.scm` does what you want, I think...

You are right @jpellegrini, and it is the reason for existence of `stklos-script`. Anyway, `stklos-script` is less general than the `stklos` command, since it doesn't accept `stklos` command's options.  The initial reason to impose the `-f` option in `stklos` was to permit scripts coming from `stdin`.

```sh
$ echo '(write (command-line))' | st -n *.c
("" "base64.c" "boolean.c" "boot.c" ...)
```

If the first parameter is the name of the script, as suggested by @lassik, we would take `base64.c` as the name of the script to run. To avoid this problem, we can now use the `--` special delimiter. The above script should be written as: 

```sh
$ echo '(write (command-line))' | st -n -- *.c
```
Note that most of the scheme interpreters (except `guile`, in some extent) and other languages such as Python
or Ruby still take `base64.c` as the name of the script to run even in the presence of the `--`. 

I think that the current commit permit to be not too different of other interpreters, while still allowing to have scripts coming for `stdin` .

**Comment by egallesio** [comment]
Now the question is: “Do we still need the `stklos-script` command?"

**Comment by lassik** [comment]
Thanks a lot for implementing the feature!

> Note that most of the scheme interpreters (except `guile`, in some extent) and other languages such as Python
> or Ruby still take `base64.c` as the name of the script to run even in the presence of the `--`.

They tend to use single dash (`-`) to read from stdin. Many classic Unix tools also do this.

```
$ echo 'puts(ARGV.inspect)' | ruby - a b c
["a", "b", "c"]

$ echo 'import sys; print(repr(sys.argv))' | python - a b c
['-', 'a', 'b', 'c']
```

The current STklos implementation of the concept has the following anomalies:

```
$ ./hello.scm -- a b -c
("./hello.scm" "a" "b" "-c")  ; Drops the "--" argument.

$ ./hello.scm a b -c
("./hello.scm" "a" "b")       ; Drops the "-c" argument.
```

**Comment by egallesio** [comment]
> They tend to use single dash (`-`) to read from stdin. Many classic Unix tools also do this.

:+1: I should have thought to use `-`  I will implement that.

The first `--` argument is always dropped by the `getopt` function. In a way it has the same role as the Unix '\\' to tell that the following arguments must be taken as parameter even if they look as option. It doesn't appear in the output.

```
$ stklos -d hello.scm arg1 -- -d arg2
("hello.scm" "arg1" "-d" "arg2")
```

The first `-d` permits to be in debug mode (level 1). The second `-d` is seen as a parameter, since it is after a `--`. Omitting the second `--`, the second `-d` would have placed the debug mode at level 2.

```
$ stklos -d hello.scm arg1 -d arg2
%guess-pathname: trying "/home/eg/.stklos/stklosrc"
%guess-pathname: trying "/home/eg/.stklos/stklosrc.so"
%guess-pathname: trying "/home/eg/.stklos/stklosrc.ostk"
...
%guess-pathname: trying "./hello.scm"
;; Loading file "./hello.scm".
("hello.scm" "arg1" "arg2")
;; File "./hello.scm" loaded.
```

About:

```
$ ./hello.scm a b -c
("./hello.scm" "a" "b")       ; Drops the "-c" argument.
```

This is “normal”. The GNU `getopt` permits to have options after the parameters (this was different with original Unixes). In this case, the `-c` is taken as an option, since it is a valid STklos option.

**Comment by lassik** [comment]
The usual approach is to separate the command line into two parts: the STklos part and the script's part, like so:

```
stklos [stklos options ...] script.scm [script options ...]
```

So `stklos -c script.scm -d` would pass `-c` to STklos but `-d` to the script.

**Comment by egallesio** [comment]
On Unix v7, the first version I used back in 1983 or 1984, it was the rule. But the glibc permits to mix parameters and options. So this is no more the case, at least on systems such as GNU/Linux. It seems that BSD derived systems are more conventional. For instance, on Linux

```
$ cat - -n
```
 
will copy  `stdin` on `stdout`, numbering the lines. On MacOs, it will copy `stdin` on `stdout` and probably declare after that it can open a file named `-n`. 

> So `stklos -c script.scm -d` would pass `-c` to STklos but `-d` to the script. 

To pass the `-d` to the script, with gnu `getopt` used by STklos, you can 

* use the `--` option: `$ stklos -c script.scm -- -d`
* use the `POSIXLY_CORRECT` environment variable : `$ POSIXLY_CORRECT=1 stklos -c script.scm  -d`


Commands such as `gcc` or `clang`, also accepts options after parameters. However, you are right, interpreters such as Python, Ruby, or Lua do `not. 

Anyway, I have already implemented your suggestion with the '-' for denoting stdin. But I'm torn between using the GNU form or staying traditional (this is just a matter to add or not a '+' in front of the last parameter of `getopt`).

Funny to see the intricacies implied by this issue :smile:

**Comment by lassik** [comment]
> Funny to see the intricacies implied by this issue 😄

Yes, command line parsing can be pretty awful :)

Commands that run subcommands tend to behave differently from commands that don't.

In `gcc foo.c -o foo` the option `-o foo` comes after the non-option argument `foo.c`. But since gcc doesn't pass any of the options to a subcommand, there is no ambiguity about who the options belong to. They all belong to gcc.

Commands like `git` pass the tail of the command line to a subcommand. For example, `git --no-pager log` passes the `--no-pager` option to the main `git` command. By contrast, `git log --no-pager` passes the option to the `log` subcommand, which fails because that subcommand does not recognize that option.

This strategy used by `git` et.al. is logical and works well in practice. I recommend it for STklos.

**Comment by egallesio** [comment]
You convinced me: option must be before the script name. The last commit should be OK.

**Comment by lassik** [comment]
Thank you very much for listening!

I built and tested the latest commit, and all of the use cases now work naturally.

> Now the question is: “Do we still need the `stklos-script` command?"

Is there anything it can do that can't be done by putting `#! /usr/bin/env stklos` at the top of a script?

**Comment by lassik** [comment]
> On Unix v7, the first version I used back in 1983 or 1984, it was the rule.

Quite impressive BTW!

**Comment by egallesio** [comment]
> Is there anything it can do that can't be done by putting `#! /usr/bin/env stklos` at the top of a script?

I don't think so, and I will do the replacement everywhere. I'll keep the `stklos-script` command for old compiled scripts, but don't use it anymore.

  

> > On Unix v7, the first version I used back in 1983 or 1984, it was the rule.
> 
> Quite impressive BTW!

It is just that I'm old (I used it on a PDP11/23 during my Master).
